File: module/storage-kv/db_lease_exp.go
package main

import (
	"encoding/binary"
	"iter"

	"github.com/pantopic/wazero-lmdb/sdk-go"
)

type dbLeaseExpImpl struct {
	db
}

func (db dbLeaseExpImpl) init(txn *lmdb.Txn) {
	db.open(txn)
}

func (db dbLeaseExpImpl) put(txn *lmdb.Txn, item lease) error {
	k := binary.BigEndian.AppendUint64(nil, item.expires)
	k = binary.AppendUvarint(k, item.id)
	return txn.Put(db.i, k, db.addChecksum(k, nil), 0)
}

func (db dbLeaseExpImpl) del(txn *lmdb.Txn, item lease) (err error) {
	key := binary.BigEndian.AppendUint64(nil, item.expires)
	key = binary.AppendUvarint(key, item.id)
	return txn.Del(db.i, key, nil)
}

func (db dbLeaseExpImpl) scan(txn *lmdb.Txn, expires uint64) iter.Seq[uint64] {
	cur, err := txn.OpenCursor(db.i)
	if err != nil {
		return nil
	}
	k, v, err := cur.Get(nil, nil, lmdb.Next)
	return func(yield func(uint64) bool) {
		defer cur.Close()
		for !lmdb.IsNotFound(err) {
			if err != nil {
				break
			}
			_, err = db.trimChecksum(k, v)
			if err != nil {
				break
			}
			if len(k) < 9 {
				break
			}
			if binary.BigEndian.Uint64(k[:8]) > expires {
				break
			}
			id, _ := binary.Uvarint(k[8:])
			if !yield(id) {
				break
			}
			k, v, err = cur.Get(nil, nil, lmdb.Next)
		}
	}
}

File: module-zig/storage-kv/db_lease_exp.zig
//! Mirrors module/storage-kv/db_lease_exp.go

const std = @import("std");
const mdb = @import("mdb");

const Db = @import("db.zig").Db;
const Lease = @import("lease.zig").Lease;
const util = @import("util.zig");

pub const DbLeaseExp = struct {
    db: Db,

    pub fn init(self: DbLeaseExp, txn: mdb.Txn) void {
        self.db.open(txn);
    }

    fn keyOf(item: Lease, buf: *[8 + util.max_varint_len]u8) []const u8 {
        std.mem.writeInt(u64, buf[0..8], item.expires, .big);
        const n = util.putUvarint(buf[8..], item.id);
        return buf[0 .. 8 + n];
    }

    pub fn put(self: DbLeaseExp, txn: mdb.Txn, item: Lease) !void {
        var kbuf: [8 + util.max_varint_len]u8 = undefined;
        const k = keyOf(item, &kbuf);
        var vbuf: [4]u8 = undefined;
        return txn.put(self.db.i, k, self.db.addChecksum(k, "", &vbuf), 0);
    }

    pub fn del(self: DbLeaseExp, txn: mdb.Txn, item: Lease) !void {
        var kbuf: [8 + util.max_varint_len]u8 = undefined;
        return txn.del(self.db.i, keyOf(item, &kbuf), "");
    }

    /// Iterates ids of leases expiring at or before `expires`.
    /// Mirrors Go's scan iterator; the cursor closes when iteration ends.
    pub fn scan(self: DbLeaseExp, txn: mdb.Txn, expires: u64) Scan {
        const cur = txn.openCursor(self.db.i) catch {
            return .{ .db = self.db, .cur = null, .expires = expires };
        };
        return .{ .db = self.db, .cur = cur, .expires = expires };
    }

    pub const Scan = struct {
        db: Db,
        cur: ?mdb.Cursor,
        expires: u64,

        pub fn next(self: *Scan) ?u64 {
            const cur = self.cur orelse return null;
            const entry = cur.get("", "", mdb.op_next) catch {
                self.close();
                return null;
            };
            _ = self.db.trimChecksum(entry.key, entry.val) catch {
                self.close();
                return null;
            };
            if (entry.key.len < 9) {
                self.close();
                return null;
            }
            if (std.mem.readInt(u64, entry.key[0..8], .big) > self.expires) {
                self.close();
                return null;
            }
            const id = util.uvarint(entry.key[8..]) orelse {
                self.close();
                return null;
            };
            return id.v;
        }

        pub fn close(self: *Scan) void {
            if (self.cur) |cur| {
                cur.close();
                self.cur = null;
            }
        }
    };
};
